### DISCUSSION TITLE
c# keep parent alive when holding reference to child

### DISCUSSION THREAD

**Discussion by roderickgreen** [swig/swig#3006] [General]
I'm adopting SWIG in a large codebase. In our use, we frequently hold onto SWIG proxy child objects without holding onto the parent. This results in memory corruption when the GC collects the parent and the child now points to freed memory.  Note that this is a more general case of #945.  

I have a fix locally to add an owner reference to all child proxy instances, and I'd like to discuss getting this merged into SWIG if we can make it general enough to cover all use cases. 

A sketch of my initial fix is to define a new macro to replace SWIG_CSBODY_PROXY, add an object cOwner argument, and hold on to a reference (only base ctor shown):
```
%define SWIG_CSBODY_PROXY_OWNER(PTRCTOR_VISIBILITY, CPTR_VISIBILITY, TYPE...)
// Proxy classes (base classes, ie, not derived classes)
%typemap(csbody) TYPE %{
  private global::System.Runtime.InteropServices.HandleRef swigCPtr;
  protected bool swigCMemOwn;
  private object swigCOwner;

  PTRCTOR_VISIBILITY $csclassname(global::System.IntPtr cPtr, bool cMemoryOwn, object cOwner) {
    swigCMemOwn = cMemoryOwn;
    swigCPtr = new global::System.Runtime.InteropServices.HandleRef(this, cPtr);
    swigCOwner = cOwner;
  }
```

Then I fumbled through the typemaps to pass `this` or `null` as the owner depending on the context.  

This is something I'd really like to contribute back to SWIG since I'm sure others will need it.  But I'm going to need some handholding to make it work in more general cases.

**Comment by ojwb** [comment]
https://sourceforge.net/p/swig/bugs/1314/ in the old tracker looks related too.

**Reply by roderickgreen** [reply]
Yeah this is the same issue, thanks. I didn't come across this report in my initial search. I haven't unpacked the patch in that report to compare the approaches yet.  I do think it's worthwhile taking the time to get a fix merged since people will continue to rediscover it, and it's not easy to discover the root cause each time.

**Comment by roderickgreen** [comment]
I'm getting up to speed on SWIG dev and can run tests now. I'm starting by collecting some failing tests in https://github.com/roderickgreen/swig/tree/csharp-gc-keepalive

**Comment by roderickgreen** [comment]
I like the approach Greg used to hold onto the owner in swigCPtr instead of a new member like I did initially. This will end up looking very close to his original patch.

**Comment by wsfulton** [comment]
I've updated the SF bug with more info. In summary, the recommended C# solution was documented (many years later), but perhaps not as well as it could have been for member variables. I've now updated this in master, which you can see in the "Memory management for accessing member variables" section, try this link: https://htmlpreview.github.io/?https://github.com/swig/swig/blob/master/Doc/Manual/CSharp.html#CSharp_memory_management_member_variables

The patch at the start of this discussion effectively implements the same thing as what the docs recommend, but requires an additional `addReference()` call instead of adding a reference in the constructor. This provides flexibility and is added in only when uses want it.

Regarding the SF patch, I'm still trying to convince myself it is safe as it replaces the object that `HandleRef` points to which is wrong as it should be pointing to the this pointer. It removes a reference that is required, but am not really sure if it is safe in all circumstances. See what you think of the documented solution.

**Comment by roderickgreen** [comment]
Hi William, I appreciate you taking the time to respond, and I apologize for not following up sooner.  

I do think the docs look good. Could we add the text `Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.` to this section in the docs as a warning of the types of errors you can see when this pattern is not followed? I searched quite a bit during my initial triage of the error didn't come across this whole section of documentation that would have immediately pointed me in the right direction.  

As for patching SWIG to avoid this in the general case, I do think it's worth continuing investigation into ways the proxy code can be generated so as to avoid the issue entirely. I don't have strong opinions about how the patch looks. My initial once-over got maybe 80% of the way there, with a long tail of tests that contained code I couldn't figure out how to generate.  Then I put it down and haven't looked at it again.  If you or another member of the team is willing to hold my hand some, I can iterate on the tests over time and try to get everything sorted. Or we can shelve this for later consideration.

**Reply by wsfulton** [reply]
Error message added in 1e45590221ff47abcbd93a4df5a0f726ee1d8f44.

I'm afraid in an all-volunteer project there is never enough time to deal with the mountains of issues in such a complex project like SWIG. However, I'm happy to review and discuss patches noting that a fundamental change like this would need to be rock solid in order to be accepted and it feels like that might be a fair bit of testing and thinking.

**Reply by roderickgreen** [reply]
Awesome, thanks again, William. If I come up with something I'm really happy with, I'll follow up to get your feedback.
